src/App.tsx

```diff
@@ -15,19 +15,25 @@ import LoginControl from "./LoginControl"
 
 import "./App.scss"
 import logoDark from "./images/logo-dark.png"
+import { connect } from "tls"
 
 interface AppState {
-  loggedIn: boolean
+  loggedIn: string
 }
 
 export default class App extends React.Component<unknown, AppState> {
   constructor(props: unknown) {
     super(props)
     this.state = {
-      loggedIn: localStorage.getItem(consts.brigadeAPITokenKey) ? true : false
+      loggedIn: ""
     }
   }
 
+  identifyUser = async () => {
+    const userId = (await getClient().authn().whoAmI()).id
+    this.setState({ loggedIn: userId})
```

Comment:
No issues ! I will create a getUsername function like you've created one for the singleton APIClient. Working on it asap.

Reply:
That would seem to be the normal way of doing things in react... however, that's going to mean, in many cases, passing this information from component to component to component. It becomes messy quickly.

I'm new to react, but I recognize this as a more general coding problem that exists in most languages, frameworks, etc. -- on both the front end and backend. Knowing _who_ the current principal is is a so-called _cross-cutting_ concern. If you pass this information around to every place it's needed, interfaces become "polluted." e.g. Once you get to the point where _most_ functions take some kind of user context as an argument, it just becomes _noise_.

If we want to keep the signal-to-noise ration high, we should stash this information somewhere globally accessible.

That's the common approach to solving this problem across most languages and frameworks I have encountered.

Reply:
@krancour Understood. I was bit aware about these things but I did not know the exact terminology for the same. According to me, In React when we want to make something accessible to multiple components and want them to re-render whenever they change, we simply use some external tools like Redux. You can check that out and then we can decide to use that or not as per the use case 😄

Reply:
I think forcing things to re-render isn't a huge concern here.

It probably hasn't been clear how I imagine user ID will be used once we have that information. Maybe putting it in context helps...

There are places currently where I want to query a user's permissions to conditionally render or not render certain UI elements. This requires first knowing who the user is.

A very good example is that if a user does not have permissions to view logs for a given project, I don't even want to show them the log tab. Or maybe I want to gray that tab out.

Note that the back-end already enforces all of this. If you don't have permissions to view logs for a given project, the back end will not serve them to you. So this is more a matter of UX here -- "don't tease me my showing me something I can't access."

So the more I think about this, perhaps we were headed in the wrong direction with this PR and overcomplicating things (my fault more so than yours).

Probably what this whole problem reduces to is that we need a globally accessible `getUser()` function that we can use to get the user ID when we need it. Then this solution reduces to simply implementing `getUser()` well. That could be something as simple as returning the user ID from local storage if it's there. If it isn't there, retrieve it, stash it in local storage, and return it.

Reply:
> There are places currently where I want to query a user's permissions to conditionally render or not render certain UI elements. This requires first knowing who the user is.

Acc to me whenever we conditionally render based on data ( whatever it is ) we must put it in state so that components comfortably update themselves whenever data change. but no issues ! We will go with localStorage.

`"don't tease me by showing me something I can't access."` Understood ! Gr8 Example 😂

> So the more I think about this, perhaps we were headed in the wrong direction with this PR and overcomplicating things (my fault more so than yours)

So Rather than overcomplicating this, we can simply create a global function for now !